File: backend/services/userService.js
const User = require('../models/User');

/**
 * Get all users (admin view)
 */
const getAllUsers = async () => {
  return User.find({}).select('-password').sort({ createdAt: -1 });
};

/**
 * Get user by ID
 */
const getUserById = async (id) => {
  const user = await User.findById(id).select('-password');
  if (!user) {
    const err = new Error('User not found.');
    err.statusCode = 404;
    throw err;
  }
  return user;
};

/**
 * Create a new user (admin only)
 */
const createUser = async ({ name, email, password, role }) => {
  const existing = await User.findOne({ email });
  if (existing) {
    const err = new Error('A user with this email already exists.');
    err.statusCode = 409;
    throw err;
  }
  return User.create({ name, email, password, role: role || 'Member' });
};

/**
 * Update a user
 */
const updateUser = async (id, updates) => {
  const allowedFields = ['name', 'email', 'role', 'isActive'];
  const filteredUpdates = Object.fromEntries(
    Object.entries(updates).filter(([k]) => allowedFields.includes(k))
  );

  const user = await User.findByIdAndUpdate(id, filteredUpdates, {
    new: true,
    runValidators: true,
  }).select('-password');

  if (!user) {
    const err = new Error('User not found.');
    err.statusCode = 404;
    throw err;
  }

  return user;
};

/**
 * Soft-delete (deactivate) a user
 */
const deleteUser = async (id, requestingUserId) => {
  if (id === requestingUserId.toString()) {
    const err = new Error('You cannot delete your own account.');
    err.statusCode = 400;
    throw err;
  }

  const user = await User.findByIdAndDelete(id);
  if (!user) {
    const err = new Error('User not found.');
    err.statusCode = 404;
    throw err;
  }

  return user;
};

module.exports = { getAllUsers, getUserById, createUser, updateUser, deleteUser };

File: flask-backend/app/services/user_service.py
from ..extensions import db
from ..models.user import User


class UserService:

    @staticmethod
    def list_users() -> list:
        users = User.query.order_by(User.created_at.desc()).all()
        return [u.to_dict() for u in users]

    @staticmethod
    def create(data: dict) -> dict:
        if User.query.filter_by(email=data["email"]).first():
            raise ValueError(f"Email '{data['email']}' is already registered")

        user = User(
            name=data["name"],
            email=data["email"],
            role=data.get("role", "member"),
        )
        user.set_password(data["password"])
        db.session.add(user)
        db.session.commit()
        return user.to_dict()

    @staticmethod
    def update(user_id: int, data: dict) -> dict:
        user = db.session.get(User, user_id)
        if not user:
            raise LookupError("User not found")

        if "name" in data:
            user.name = data["name"]
        if "email" in data:
            user.email = data["email"]
        if "role" in data:
            user.role = data["role"]
        if "is_active" in data:
            user.is_active = data["is_active"]
        if "password" in data:
            user.set_password(data["password"])

        db.session.commit()
        return user.to_dict()

    @staticmethod
    def delete(user_id: int) -> None:
        user = db.session.get(User, user_id)
        if not user:
            raise LookupError("User not found")
        db.session.delete(user)
        db.session.commit()
